Add regression test to volunteer op#clear_lat_lng callback

Maroo-b opened the pull request:
Addresses: https://www.pivotaltracker.com/story/show/129203805
- check that callback keeps remote volunteer op longitude and latitude
  coordinates

- refactor `clear_lat_lng` callback tests

tansaku commented:
@Maroo-b thanks for this

Is there any chance we can connect it to a cucumber feature test?  I'm slightly unclear about how the circumstance this checks for actually arises - can we specify that in terms of a feature test?

Maroo-b commented:
@tansaku I added a step to the import doit feature (https://github.com/AgileVentures/LocalSupport/pull/354/files#diff-28c62bf658c58fa72820964520efda15R10) to check that all imported volunteer op from doit have longitude and latitude coordinates.

It seems now that we have a base regression tests to prevent the same bug from occurring.

Maroo-b commented:
@tansaku we have the login feature fail again, can you restart the build ?

tansaku commented:
@Maroo-b build re-started (and I created a ticket related to those build problems: https://www.pivotaltracker.com/story/show/129522757) - thanks for adding the additional test, but I'm still feeling I want a cucumber test to really understand the big picture here.  Should I be referring to the cucumber test in #335 to understand what's going on here?

Should we not pull that cucumber test into this PR, so that they can be reviewed in concert, or do you think that's not appropriate here?

Many thanks in advance

Maroo-b commented:
@tansaku you are right, I'll try to summarize the production bug and investigations note:

**Bug description steps:**
When a user visits volunteer op index page
Then he gets a server error page 

**Investigation Notes:**
- The function responsible for the bug in volunteer op index page was the map marker, in this scenario it was trying to display a marker for a volunteer record without coordinates ( that was the raised exception)

- Next step we queried the volunteer op DB to fetch records without coordinates we noted that all of them are from 'doit' source. As a conclusion, the import volunteer op task was the first to check.

- After playing around with the console, it was the clear_lat_lng callback from volunteer op model which was responsible for this behavior (it was triggered against all sources local and doit)

**Fix implemented:**

- Make clear_lat_lng callback triggers only with local source ( fix added with unit test support at this level)

I hope that I made the issue and the modifications made clear, for the moment the I added a step to import task feature test to check if the imported volunteer op tasks have coordinates in the DB.
Do you think about other alternatives ?

tansaku commented:
I'll review and come back to you on this tomorrow @Maroo-b

tansaku commented:
Thanks for the detailed write up here @Maroo-b - I'm a little confused though, because I don't remember this particular bug.  At least, I thought the issue we were having was that while the rest of the volunteer ops were being displayed, none of the do-it ops was being displayed, i.e. the import operation was failing (#335) the PR to fix that had some new acceptance tests which we don't have here.  Is this PR not related to #335 ? Do we not need to pull those acceptance tests in here?

Or am I remembering incorrectly?  Perhaps this is just a different bug and I am confusing the two, in which case my sincere apologies.

However, if this bug caused a server error when the user visits the volunteer ops page, then it should be covered by an acceptance test. Or is it covered by an existing cucumber acceptance test?  

What I'd really like is to be able to run that acceptance test and see the failure and then the fix.  Can you point me to that existing acceptance test, or tell me if there's a reason why we can't create an acceptance test for this issue?  Or maybe you think it's not needed in this case?

Does that make sense?

Maroo-b commented:
@tansaku I agree that it seems confusing but I'll try to make it simple as possible.


So for the first question:
>  the import operation was failing (#335) the PR to fix that had some new acceptance tests which we don't have here. Is this PR not related to #335 ? Do we not need to pull those acceptance tests in 

The 2 PR are related  #335 was intended to fix the production bug ( but after rolling back to fix travis-ci failing build) the fix was included in another ticket https://github.com/AgileVentures/LocalSupport/pull/343 but without any tests !!!

So I submitted this PR to add tests to prevent facing the same bug in production.

So to reproduce the bug technically you can delete manually longitude and latitude coordinates for **doit** volunteer op source and visit volunteer op index page

So the question can we reproduce this issue as acceptance test ? 
As first answer the regression test should be on volunteer op index page ( where we had the issue) right ? But technically it's related to import doit task tests (even if it's really **should be** the volunteer op model). So that's why I added the additional step in the import tasks acceptance test to be sure that all imported op have coordinates (this test fails if the callback `clear_lat_lng` from Volunteer op model is triggered to all records not only local) 

So if we return our attention now to the model Volunteer op, I added unit tests to assure that the clearing coordinates callback don't affect remote sources.

I hope that I was clear :)

tansaku commented:
> As first answer the regression test should be on volunteer op index page ( where we had the issue) right?

Not necessarily - I'd say that the key thing is an acceptance test that replicates the issue as it occurred in the system, independently of how many places that has to touch, whether that's services or controllers or what have you.

I find it difficult to evaluate the individual regression test at the level of the controller without the presence of the acceptance test that precisely defines the nature of the problem.  I really appreciate you taking the trouble to explain the details in English, but cucumber/capybara/rspec is preferrable. Do you know what I mean?

Maroo-b commented:
@tansaku what do you suggest as acceptance test in this case ?

Maroo-b commented:
The regression test I thought to use:

As admin
I run the import doit service
Then should be x doit volunteer op stored
**And all imported volunteer op have coordinates**

what do you think about this regression test ? and do you suggest another one ?

tansaku commented:
I would call that an "acceptance" test rather than a regression test, but that's fine if it's difficult to check the display of the coordinates in the map, but we have a test like that don't we?

```gherkin  Background:
    Given that the doit_volunteer_opportunities flag is enabled
    And I run the import doit service
    And cookies are approved

  Scenario: See a map of current doit volunteer opportunities
    And I visit the volunteer opportunities page
    And I should see 13 markers in the map
```
in `do_it_volunteer_opportunities.feature` and if I remember this one was the one that was failing in that branch with the failing tests, and passed when I made the fix, so identifying this is helpful.   Your acceptance that you describe would be another good check.  With those in mind I can start to assess the lower level tests more effectively :-)

tansaku commented:
and in fact you have a reasonable acceptance test update in there in `features/devops/import_doit.feature` - okay - sorry that took a while for me to get my head round - I can evaluate properly now I think - just leave this one with me and I'll get it merged in soon now I think :-)

tansaku commented:
@Maroo-b sorry for not completing the review here yet - could you update to develop for me? thanks - will prioritize this for Monday

Maroo-b commented:
@tansaku I updated the branch, we having a failing build due to JS again :/

tansaku commented:
thanks @Maroo-b - have restarted the build - hopefully @johnnymo87 's PR might fix this ...

tansaku commented:
@Maroo-b - I just merged @johnnymo87 rerunner - want to pull that in to see if it fixes the build - also was that rubocop-git tool I mentioned helpful? https://github.com/m4i/rubocop-git

Can you pinpoint for me which hound, or codeclimate, issues that don't make sense to you?

Maroo-b commented:
@tansaku I'll do, but the hounds and codeclimate issues are on this PR https://github.com/AgileVentures/LocalSupport/pull/358 :)

Maroo-b commented:
@tansaku PR updated.

tansaku commented:
cool @Maroo-b - I just refactored a little - take a look.

also I wonder if we might make a custom matcher to check for non-nil lat/lng - little bit of a luxury, but might be fun:

http://www.relishapp.com/rspec/rspec-expectations/v/3-5/docs/custom-matchers/define-a-custom-matcher#define-a-matcher-with-default-messages

tansaku commented:
@johnnymo87 example of the re-run showing a coverage drop as I guessed might happen: 

![](https://www.dropbox.com/s/c1kmt9cqlst7p4o/Screenshot%202016-09-21%2013.52.21.png?dl=1)

Maroo-b commented:
@tansaku I added a custom matcher for the coordinates.

tansaku commented on spec/support/matchers/coordinates_matcher.rb:
```diff
@@ -0,0 +1,10 @@
+RSpec::Matchers.define :have_coordinates do |expected|
```
I'd just name this `_expected` here to indicate that we don't use it

tansaku commented:
fantastic work @Maroo-b !
